### PR TITLE
[tests-only] Tags

### PR DESCRIPTION
**Pull request by kobergj**
This adds the possibility to add and remove tags to files and folders. After adding a tag it will reported through the PROPFIND response

### Setting tags
set a tag using the new webdav tags api:
```bash
curl --insecure -X PUT -u admin:admin "https://localhost:9200/remote.php/dav/tags/1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1?tags=coolstuff"
```
set multiple tags at once by providing a comma separated list:
```bash
curl --insecure -X PUT -u admin:admin "https://localhost:9200/remote.php/dav/tags/1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1?tags=coolstuff,nicetag"
```
Providing an empty tags list or a list containing no new tags will result in a `BadRequest`

### Deleting Tags
delete tags using the new webdav tags api:
```bash
curl --insecure -X DELETE -u admin:admin "https://localhost:9200/remote.php/dav/tags/1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1?tags=coolstuff"
```
delete multiple tags at once by providing a comma separated list:
```bash
curl --insecure -X DELETE -u admin:admin "https://localhost:9200/remote.php/dav/tags/1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1?tags=coolstuff,nicetag"
```
Providing an empty tags list or a list containing no existing tags will result in a `BadRequest`

### Read Tags
tags will be reported in the `PROPFIND` response in the `oc:tags` prop. They will be ordered by time of creation, newest first. 
```bash
curl --insecure -X PROPFIND -u admin:admin "https://localhost:9200/remote.php/dav/spaces/1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1"

<?xml version="1.0"?>
<d:multistatus xmlns:s="http://sabredav.org/ns" xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/remote.php/dav/spaces/1284d238-aa92-42ce-bdc4-0b0000009157$4ce9c71e-62f3-4765-8d32-1a9139f7869c%211dc64a92-3d0f-4050-872a-6c035ce252e1</d:href>
    <d:propstat>
      <d:prop>
        <oc:id>1284d238-aa92-42ce-bdc4-0b0000009157$4ce9c71e-62f3-4765-8d32-1a9139f7869c!1dc64a92-3d0f-4050-872a-6c035ce252e1</oc:id>
        <oc:fileid>1284d238-aa92-42ce-bdc4-0b0000009157$4ce9c71e-62f3-4765-8d32-1a9139f7869c!1dc64a92-3d0f-4050-872a-6c035ce252e1</oc:fileid>
        <oc:spaceid>4ce9c71e-62f3-4765-8d32-1a9139f7869c</oc:spaceid>
        <d:getetag>"35c3ec1cff804bb40c239849c724523e"</d:getetag>
        <oc:permissions>RDNVW</oc:permissions>
        <d:resourcetype/>
        <d:getcontentlength>3343173</d:getcontentlength>
        <d:getcontenttype>application/pdf</d:getcontenttype>
        <d:getlastmodified>Tue, 26 Jul 2022 10:49:30 GMT</d:getlastmodified>
        <oc:checksums>
          <oc:checksum>SHA1:f157ec1d2e25834b80c7c0657a6e406d3eec42f9 MD5:d53bc1e25c0c495e0db414523940efeb ADLER32:881bdd93</oc:checksum>
        </oc:checksums>
        <oc:tags>coolstuff,nicetag</oc:tags>
        <oc:favorite>0</oc:favorite>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

Note: All examples use an item with the resourceid `1284d238-aa92-42ce-bdc4-0b0000009157\$4ce9c71e-62f3-4765-8d32-1a9139f7869c\!1dc64a92-3d0f-4050-872a-6c035ce252e1` (with escaped $ and ! characters because of the terminal) replace that with your rid

### CODE REVIEW

**Comment by phil-davis** [comment]
Note: tags in oC10 have various attributes, so they are reported in PROPFIND requests with more structure. Here is an example on my local test system:
```
$ cat tag.xml 
<?xml version="1.0" encoding="utf-8" ?>
<a:propfind xmlns:a="DAV:" xmlns:oc="http://owncloud.org/ns">
  <a:prop>
    <oc:display-name/>
    <oc:user-visible/>
    <oc:user-assignable/>
    <oc:id/>
  </a:prop>
</a:propfind>
```

```
$ curl -u aaaa:aaaa -X PROPFIND -H "Content-Type: text/xml" --data-binary '@tag.xml' http://172.17.0.1:8080/remote.php/dav/systemtags-relations/files/2147507556 | xmllint --format -
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1235    0  1010  100   225   7952   1771 --:--:-- --:--:-- --:--:--  9724
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/remote.php/dav/systemtags-relations/files/2147507556/</d:href>
    <d:propstat>
      <d:prop>
        <oc:display-name/>
        <oc:user-visible/>
        <oc:user-assignable/>
        <oc:id/>
      </d:prop>
      <d:status>HTTP/1.1 404 Not Found</d:status>
    </d:propstat>
  </d:response>
  <d:response>
    <d:href>/remote.php/dav/systemtags-relations/files/2147507556/3</d:href>
    <d:propstat>
      <d:prop>
        <oc:display-name>abc</oc:display-name>
        <oc:user-visible>true</oc:user-visible>
        <oc:user-assignable>true</oc:user-assignable>
        <oc:id>3</oc:id>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
  <d:response>
    <d:href>/remote.php/dav/systemtags-relations/files/2147507556/29</d:href>
    <d:propstat>
      <d:prop>
        <oc:display-name>def</oc:display-name>
        <oc:user-visible>true</oc:user-visible>
        <oc:user-assignable>true</oc:user-assignable>
        <oc:id>29</oc:id>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```

There are `user-visible` and `user-assignable` attributes.

If tags will have more structure/metadata/attributes in the future, then maybe PROPFIND should return a structured list of XML elements inside `<oc:tags>` rather than just a comma-separated list. That way, in future, extra "tag metadata" can be easily included.

**Comment by fschade** [comment]
@phil-davis, thanks, we've decided to keep it as it is now, we can change it later if needed.

File: pkg/tags/tags.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	for _, tag := range tags {
+		t.t = append(t.t, tag)
+		t.exists[tag] = true
+		t.numtags++
+	}
+	return t
+}
+
+// AddString appends the the new tags and returns true if at least one was appended
+func (t *Tags) AddString(s string) bool {
```
Comment (fschade) [inline pkg/tags/tags.go:54]: is something else planned than string? why not just `Add`?

File: pkg/tags/tags.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	t.t = append(tags, t.t...)
+	return len(tags) > 0
+}
+
+// RemoveString removes the the tags and returns true if at least one was removed
+func (t *Tags) RemoveString(s string) bool {
+	var removed bool
+	for _, tag := range strings.Split(s, t.sep) {
+		if !t.exists[tag] {
+			// should this be reported?
```
Comment (fschade) [inline pkg/tags/tags.go:80]: i say no :)

File: pkg/tags/tags_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+package tags
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAddTags(t *testing.T) {
+	testCases := []struct {
```
Comment (fschade) [inline pkg/tags/tags_test.go:28]: i woul like to see a test which is more irregular, `a,     ,, b, xf, !,`

**Review by fschade — APPROVED** [review]
just some comments, nothing big and totally optional, LGTM, THX!

File: pkg/tags/tags.go (hunk 1 above)
Comment (kobergj) [reply pkg/tags/tags.go:54]: If I would call `tags.Add(s)` I would expect `s` being one single tag. This is different here. `s` is a comma seperated list of tags. Maybe `AddMultiple`?

Comment (fschade) [reply pkg/tags/tags.go:54]: YES :)

Comment (kobergj) [reply pkg/tags/tags.go]: Did go with `AddList` `RemoveList` `FromList` because thats what it does. It adds a list of tags

**Pull request merged**